arch/x86/kernel/slaunch.c

```diff
@@ -27,13 +27,72 @@
 #include <asm/e820/api.h>
 #include <asm/bootparam.h>
 #include <asm/setup.h>
+#include <linux/fs.h>
 #include <linux/slaunch.h>
 
+#define MSG_BUFFER_LEN 20
+#define MSG_ERROR "Mapping Error\n"
+
 #define PREFIX	"SLAUNCH: "
 
+#define DECLARE_PUB_READ(reg_name, reg_offset, reg_size)					\
+static ssize_t reg_name##_read(struct file *flip, char *buffer, size_t len, loff_t *offset) {	\
+	txt_info_to_buffer(reg_offset, reg_size);						\
+	return simple_read_from_buffer(buffer, len, offset, &msg_buffer, MSG_BUFFER_LEN);	\
```

Comment:
Why are you bouncing through a static buffer instead of just writing the info directly into the destination buffer?

Reply:
In my latest commit I changed msg_buffer to be a local variable. I tried to get rid of it entirely but without it I don't believe there is any way to properly format the data with sprintf.

Reply:
Thinking about it, all of these are individual values not a long buffer that could require multiple reads from user space. You should just set them up using DEVICE_ATTR and then you can register them as a device attr group which will automatically take care of the subdirectory concern.

Reply:
hm, would I be instead exposing the registers over devfs if I did that? I'm not sure how to use DEVICE_ATTR or .show with securityfs.

Reply:
Ack, I got crossed with sysfs.